commands/migratelxc.go

```diff
@@ -0,0 +1,346 @@
+			names := containerNames[container]
+			if lxdContainers[names.oldName] != nil {
+				// migrated but not yet renamed
+				logger.Debugf(
+					"LXC container %q already migrated to LXD (%q)",
+					container.Id(), names.oldName,
+				)
+				continue
+			}
+			if lxdContainers[names.newName] != nil {
+				// migrated and renamed
+				logger.Debugf(
+					"LXC container %q already migrated to LXD (%q)",
+					container.Id(), names.newName,
+				)
+				continue
+			}
+			nonMigrated = append(nonMigrated, container)
+		}
+		if len(nonMigrated) > 0 {
+			lxcToMigrateByHost[host] = nonMigrated
+		}
+	}
+	return lxcToMigrateByHost
+}
+
+// stopLXCContainers stops all of the LXC containers.
+func stopLXCContainers(lxcByHost map[*state.Machine][]*state.Machine) error {
+	var group errgroup.Group
+	for host, containers := range lxcByHost {
+		for _, container := range containers {
+			logger.Debugf("stopping LXC container %q", container.Id())
+			host, container := host, container // copy for closure
+			group.Go(func() error {
+				return errors.Annotatef(
+					StopLXCContainer(container, host),
+					"stopping LXC container %q",
+					container.Id(),
+				)
+			})
+		}
+	}
+	return group.Wait()
+}
+
+// migrateLXCContainers migrates all of the LXC containers to LXD.
+func migrateLXCContainers(lxcByHost map[*state.Machine][]*state.Machine) error {
+	opts := MigrateLXCOptions{
+		// TODO(axw) option to copy rootfs?
+		MoveRootfs: true,
+	}
+	var group errgroup.Group
+	for host, containers := range lxcByHost {
+		containerNames := make([]string, len(containers))
+		for i, container := range containers {
+			containerNames[i] = container.Id()
+		}
+		logger.Debugf("migrating LXC containers: %s", strings.Join(containerNames, ", "))
+		host, containers := host, containers // copy for closure
+		group.Go(func() error {
+			return errors.Annotatef(
+				MigrateLXC(containers, host, opts),
+				"migrating LXC containers: %s", strings.Join(containerNames, ", "),
+			)
+		})
+	}
+	return group.Wait()
+}
+
+// renameLXDContainers renames all of the LXD containers to the new name,
+// if they aren't already named as such.
+func renameLXDContainers(
+	st *state.State,
+	lxcByHost map[*state.Machine][]*state.Machine,
+	lxdByHost map[*state.Machine]map[string]*api.Container,
+	containerNames map[*state.Machine]containerNames,
+) error {
+	environUUID := st.EnvironUUID()
+	renameLXDContainer := func(newName, oldName string, host *state.Machine) error {
+		if err := SetLXDContainerConfig(oldName, "user.juju-model", environUUID, host); err != nil {
+			return errors.Trace(err)
+		}
+		return RenameLXDContainer(newName, oldName, host)
+	}
+	var group errgroup.Group
+	for host, containers := range lxcByHost {
+		// lxcByHost contains all of the containers recorded in state,
+		// whether or not they've been migrated. We filter out the
+		// ones that were already migrated and renamed in a previous
+		// session.
+		lxdContainers := lxdByHost[host]
```

Comment:
At this point you'll have migrated some more LXCs to LXD, but they won't be in `lxdByHost`, will they?

Oh, I think I understand - you're going through them with the same logic as you did in `getLXCContainersToMigrate` and renaming any container (even if it wasn't originally in `lxdByHost`) that isn't already in there with its new name.

I found this confusing, so I started to type up an alternative way of doing it by returning the LXD containers needing renames found in `getLXCContainersToMigrate` and then adding more to them from `migrateLXCContainers` but by the time I was half done I convinced myself it was worse. So I'm happy with this.